File: r/lib/logging.R
# r/lib/logging.R
make_log_row <- function(...) {
  data.frame(..., stringsAsFactors = FALSE)
}

write_log_and_status <- function(row_df, log_path, status_path) {
  # keep your schema-rotation behavior or remove it; no behavior change for now
  if (file.exists(log_path)) {
    header <- names(read.csv(log_path, nrows = 0, stringsAsFactors = FALSE))
    if (!identical(header, names(row_df))) {
      rotated <- sub(
        "\\.csv$",
        paste0("_old_", format(Sys.time(), "%Y%m%d_%H%M%S"), ".csv"),
        log_path
      )
      file.rename(log_path, rotated)
      cat("Log schema changed. Rotated old log to:", rotated, "\n")
    }
  }

  if (file.exists(log_path)) {
    write.table(
      row_df,
      log_path,
      sep = ",",
      row.names = FALSE,
      col.names = FALSE,
      append = TRUE
    )
  } else {
    write.csv(row_df, log_path, row.names = FALSE)
  }

  write.csv(row_df, status_path, row.names = FALSE)
}

File: src/pbp/logging.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import datetime
from pathlib import Path

import pandas as pd


@dataclass(frozen=True)
class LogRow:
    refreshed_at: str
    season: int | None = None
    week: int | None = None
    game_ids: str | None = None
    rows_in: int | None = None
    rows_scoring: int | None = None
    rows_out: int | None = None
    status: str | None = None
    detail: str | None = None


def write_log_and_status(row: LogRow, log_path: Path, status_path: Path) -> None:
    """
    Python port of r/lib/logging.R::write_log_and_status(). :contentReference[oaicite:11]{index=11}
    """
    row_df = pd.DataFrame([asdict(row)])

    # Schema-change rotation (same intent as R)
    if log_path.exists():
        try:
            header = list(pd.read_csv(log_path, nrows=0).columns)
        except Exception:
            header = []
        if header and header != list(row_df.columns):
            rotated = log_path.with_name(
                log_path.name.replace(
                    ".csv",
                    f"_old_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv",
                )
            )
            log_path.rename(rotated)

    # Append or create
    if log_path.exists():
        row_df.to_csv(log_path, index=False, header=False, mode="a")
    else:
        row_df.to_csv(log_path, index=False)

    # Status is always last row only
    row_df.to_csv(status_path, index=False)
